lib/rails_state_machine/state_machine.rb

```diff
@@ -138,5 +148,13 @@ def define_model_methods
         end
       end
     end
+
+    def prefix_for_constant_definition
+      "#{@prefix.upcase}_" if @prefix.present?
+    end
+
+    def prefix_for_method_definition
+      "#{@prefix}_" if @prefix.present?
+    end
```

Comment:
I like that you store the values of those methods in instance variables directly in `initialize`. I had previously thought of doing `@prefix_for_constant_definition ||= ...`, but it's perfectly fine to just assign the value in `initialize`. :+1: 

However, methods with the same name as instance variables should not do much more than return the value from the instance variable, maybe with additional small logic like fallbacks for `nil` values. If your method computes a value, you should prefix it with `build_`, `calculate_`, or similar.
The reasoning is that it's at least unexpected if `@foo` and `foo` behave (much) differently and we can almost always avoid that confusion.

Since calculating those values is a single line each, I suggest moving that into `initialize`. You can then just define `attr_reader`s for them and use those readers instead of the instance variables in the `define_*` methods.

So:

```ruby
    def initialize(model, state_attribute, prefix: '')
      # ...
      @prefix_for_constant_definition = "#{@prefix.upcase}_" if @prefix.present?
      @prefix_for_method_definition = "#{@prefix.downcase}_" if @prefix.present?
      # ...
    end

    attr_reader :prefix_for_constant_definition, :prefix_for_method_definition
```

(I added a `downcase` for the method name prefix because I think it makes sense. If you don't like it, leave it out.)